### PR TITLE
StreamingDataset: Add intra node shuffling to accelerate second epoch

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR improves the full shuffle strategy of the StreamingDataset.

Previously, the shuffling logic would re-assign all the chunks to all ranks for every epoch e.g it would be unlikely for a given rank to find the chunk within its cache and would re-download it causing slow down.

Now, for the epochs higher than 1, we are:
1. collect the ordered chunks associated to the first epoch for each node during epoch 1 
2. shuffle them within each node based on the current epoch number and re-flatten the chunk list
3. re-perform the chunk association to the ranks based on the new order

Fixes #\<issue_number>

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--19296.org.readthedocs.build/en/19296/

cc @borda

### CODE REVIEW

File: src/lightning/data/streaming/dataset.py [hunk 1]
```diff
@@ -97,13 +97,11 @@ def __init__(
         self._state_dict: Optional[Dict[str, Any]] = None
 
     def _create_cache(self, worker_env: _WorkerEnv) -> Cache:
-        env = Environment(dist_env=self.distributed_env, worker_env=worker_env)
-
         if _should_replace_path(self.input_dir.path):
             # FIXME: Remove the `shard_rank` from the cache_path to enable reloading chunks for the second epoch
             # without paying the cost of re-download
```
Comment (awaelchli) [inline src/lightning/data/streaming/dataset.py:102]: what about this comment?

File: src/lightning/data/streaming/dataset.py [hunk 2]
```diff
@@ -368,13 +366,17 @@ def _validate_state_dict(self) -> None:
             )
 
 
-def _try_create_cache_dir(input_dir: str, shard_rank: int = 0) -> Optional[str]:
+def _try_create_cache_dir(input_dir: str, shard_rank: Optional[int] = None) -> Optional[str]:
```
Comment (awaelchli) [inline src/lightning/data/streaming/dataset.py:369]: But now the shard_rank argument is useless, we don't use it anywhere else

File: src/lightning/data/streaming/dataset.py [hunk 3]
```diff
@@ -97,13 +97,11 @@ def __init__(
         self._state_dict: Optional[Dict[str, Any]] = None
 
     def _create_cache(self, worker_env: _WorkerEnv) -> Cache:
-        env = Environment(dist_env=self.distributed_env, worker_env=worker_env)
-
         if _should_replace_path(self.input_dir.path):
             # FIXME: Remove the `shard_rank` from the cache_path to enable reloading chunks for the second epoch
             # without paying the cost of re-download
             cache_path = _try_create_cache_dir(
-                input_dir=self.input_dir.path if self.input_dir.path else self.input_dir.url, shard_rank=env.shard_rank
```
Comment (awaelchli) [inline src/lightning/data/streaming/dataset.py:106]: If we don't pass this shard rank, how will you prevent race conditions when workers download the same file (when they have indices from the same chunk)?

File: src/lightning/data/streaming/downloader.py [hunk 4]
```diff
@@ -54,17 +54,23 @@ def download_file(self, remote_filepath: str, local_filepath: str) -> None:
...
-            local_filepath,
+            local_filepath + "_tmp",
             ExtraArgs=extra_args,
             Config=TransferConfig(use_threads=False),
         )
 
+        if not os.path.exists(local_filepath):
+            shutil.move(local_filepath + "_tmp", local_filepath)
+        else:
+            shutil.rmtree(local_filepath + "_tmp")
```
Comment (awaelchli) [inline src/lightning/data/streaming/downloader.py:65]: See comment above, I'm pretty sure since now the shard rank is removed from the directory name, there will be a rare race condition here where multiple ranks will check and move/delete.

File: tests/tests_data/streaming/test_shuffle.py
Comment (awaelchli) [inline tests/tests_data/streaming/test_shuffle.py:1]: Shouldn't this file test the FullShuffle and NoShuffle classes? I propose to add test on master first, then we can be confident that the changes in this PR don't have unintended side effects with the existing tests?

File: src/lightning/data/utilities/env.py [hunk 5]
```diff
@@ -37,7 +38,10 @@ def detect(cls) -> "_DistributedEnv":
         if world_size is None or world_size == -1:
             world_size = 1
 
-        return cls(world_size=world_size, global_rank=global_rank)
+        # TODO: Add support for other accelerators
+        num_nodes = (world_size // torch.cuda.device_count()) if torch.cuda.is_available() else 1
+
```
Comment (awaelchli) [inline src/lightning/data/utilities/env.py:43]: This is not very robust. This can lead to silent misbehavior. At the very least we should error and not just round down to the nearest integer. 

But moreover, let's read the num nodes from the environment variable, not the torch device count?

**Review by awaelchli — CHANGES_REQUESTED** [review]

File: src/lightning/data/streaming/dataset.py (hunk 1 above)
Comment (tchaton) [reply src/lightning/data/streaming/dataset.py:102]: removed.

File: src/lightning/data/streaming/dataset.py (hunk 2 above)
Comment (tchaton) [reply src/lightning/data/streaming/dataset.py:369]: removed.

File: src/lightning/data/streaming/dataset.py (hunk 3 above)
Comment (tchaton) [reply src/lightning/data/streaming/dataset.py:106]: We are using filelock to synchronise the processes.

File: src/lightning/data/streaming/downloader.py (hunk 4 above)
Comment (tchaton) [reply src/lightning/data/streaming/downloader.py]: I added filelock to prevent it. It should be fine now.

File: tests/tests_data/streaming/test_shuffle.py
Comment (tchaton) [reply tests/tests_data/streaming/test_shuffle.py:1]: The tests are in the data processor file. I need to clean up the tests.

File: src/lightning/data/utilities/env.py (hunk 5 above)
Comment (tchaton) [reply src/lightning/data/utilities/env.py:43]: I added a divisible assertion to avoid errors. I think this is enough now as it is well defined on the platform.

**Review by awaelchli — APPROVED** [review]

File: src/lightning/data/streaming/combined.py [hunk 6]
```diff
@@ -46,7 +46,7 @@ def __init__(
 
     def __len__(self) -> int:
         assert self._weights
-        return int(sum(w * len(d) for w, d in zip(self._weights, self._datasets)))
+        return int(min([1 / w * len(d) for w, d in zip(self._weights, self._datasets) if w > 0]))
```
Comment (awaelchli) [inline src/lightning/data/streaming/combined.py:49]: why is this correct? Shouldn't the length correspond to what the iterator is returning?

Comment (tchaton) [reply src/lightning/data/streaming/combined.py:49]: This is an estimation due to the sampling. But the length is the length of the smallest dataset multiplied by the inverse of its weight.

Example. 2 datasets of len 10, 100 and weights of 0.5, 0.5.

At most, we can get 10 * 2 = 20 elements. 10 elements from dataset 1 and 10 from dataset 2.

File: src/lightning/data/streaming/shuffle.py [hunk 7]
```diff
@@ -97,55 +97,102 @@ def get_chunks_and_intervals_per_ranks(self, distributed_env: _DistributedEnv, c
         # 2. Shuffle them
         indexes = range(len(chunk_intervals))
 
-        # FIXME: Shuffling should be done only within the nodes to benefit
-        # from cache if the dataset doesn't fit on the node.
-        shuffled_indexes = np.random.RandomState(seed=self.seed + current_epoch).permutation(indexes)
-        shuffled_chunk_intervals = np.asarray(chunk_intervals)[shuffled_indexes]
+        seed_shift = 0 if distributed_env.num_nodes > 1 else current_epoch
```
Comment (justusschock) [inline src/lightning/data/streaming/shuffle.py:100]: do we do it this way so we don't have inter-node shuffling? Otherwise I'd expect the seed_shift to be nonzero for multiple nodes as well.

Comment (justusschock) [inline src/lightning/data/streaming/shuffle.py:100]: could you add some comments here so that it's clearer why we do these individual things?

**Review by justusschock — APPROVED** [review]

File: src/lightning/data/streaming/shuffle.py (hunk 7 above)
Comment (tchaton) [reply src/lightning/data/streaming/shuffle.py:100]: Done.

Comment (tchaton) [reply src/lightning/data/streaming/shuffle.py:100]: done
